Tenant Token (#89)
meilisearch/specifications: text/0089-tenant-token.md

### II. Motivation

`Tenant tokens` are introduced to solve multi-tenant use-cases. By providing the management of API Keys, we remove one of the last major deal-breakers that makes users not choose MeiliSearch as a solution despite all our advantages.

Users regularly request Multi-Tenant indexes. Users today need to set up workarounds to meet this need. Some of them implement reverse-proxy or managed authentication systems like Hasura or Kong to filter what can and cannot be read. Others decide to use server code as a facade to implement the access restriction logic before requesting MeiliSearch. It is difficult to maintain, less efficient, and requires necessary skills that not everyone has.

### III. Glossary

| Term               | Definition |
|--------------------|------------|
| Master Key         | This is the master key that allows the creation of other API keys. The master key is defined by the user when launching MeiliSearch. |
| API Key            | API keys are stored and managed from the endpoint `/keys` by the master key holder. These are the keys used by the technical teams to interact with MeiliSearch at the level of the frontend/backend code. |

> | Tenant Token     | These tokens are not stored and managed by a MeiliSearch instance. They are generated for each end-user by the backend code from a MeiliSearch API Key. They are used by the end-users to only search the documents belonging to them. |

Reviewer: ```suggestion
| Tenant Token     | These tokens are not stored and managed by a MeiliSearch instance. They are generated for each end-user by the backend code from a MeiliSearch API Key. 
They are used by end-users to search only for documents to which the tenant token gives them access. |
```
Did the documents really belong to the end-users or do they just have access to them? Not sure it's better but it still confusing for me.

Author: Yes, the documents belong to them or in any case should be accessible only to them. By adding the tenant tokens it allows us to overcome two pain-full workarounds to improve the quality of life of a MeiliSearch user. Here is what they have to do to be sure that a document is really accessible to one tenant.

1 - Make an index by tenant (user, company, ...), which can be difficult to manage.
2 - To have a single index by adding a proxy that will inject the forced filter from its backend which forces to have network round trips that will become useless with the tenant token once generated since it will be able to directly exchange with MeiliSearch.

Reviewer: Thanks for the clarification 😊

> | Multi-Tenancy      | By multi-tenancy, we mean that an end-user only accesses data belonging to him within an index shared with other end-users. |

Reviewer 2: I'm missing a definition of what we mean by tenant. Is every end-user a tenant?

Author: I've added a definition in the specifcation. Let me know if it's clearer!

### V. `Tenant Token` Explanations

#### Summary Key Points

- `Tenant tokens` are generated from a MeiliSearch parent `API key` on the user's backend code. They are meant to resolve multi-tenancy by restricting access to data within an index according to the criteria chosen by the team managing the MeiliSearch instance.
- `Tenant tokens` cannot be less restrictive than their respective parent `API key` and can only be used for searching with a predefined forced filter field.
- `Tenant tokens` are not stored and thus not retrievable on MeiliSearch. This is why we highly advise setting an expiration time on them for security reasons.

#### Solving Multi-Tenancy with `Tenant tokens`

![](https://i.imgur.com/J4jVe1n.png)

Let's say that `Mark` is a developer for a SaaS platform. He would like to ensure that end-user can only access their documents at search time. **His database contains many users and he hopes to have many more in the future.**

When a user registers, the backend code generates a `Tenant token` specifically for that end-user so they can only access their documents.

> The `filter` parameter is set to restrict the search for documents having an `user_id` attribute. This `filter` parameter is contained in the `Tenant token` payload and cannot be modified during the search by the end-user making the request. `filter` can be made of any valid filters. e.g. `user_id = 10 and category = Romantic`

Reviewer 2: I understand the scenario for this spec is to restrict searches so an end-user can only search on documents they have access to, but are there other use-cases for tenant tokens?

Author: Yes, another possible use-case in the future could be to restrict tokens in terms of rate-limiting for example. That said, the tenants tokens for the moment are generated only for search, it's something we could extend to other actions in the future but in my opinion in the current state of discovery, there is little chance.

This `Tenant token` is generated from a parent `API key` used to cipher the `Tenant token`. On the MeiliSearch side, the payload of the `tenant token` defines what data the user is allowed to retrieve, in order words, MeiliSearch grants authorization based on the permission defined in the Tenant token payload.

---

> #### Generating a `Tenant token`

Reviewer 2: So, the ideal workflow is not immediately clear to me. I'm assuming the code below is front-end due to it being javascript, but of course correct me if I'm wrong.

After a developer has generated a token from a set of restrictions and an API key, what do they do with? I'm assuming they must send it to MeiliSearch. Is there a new search parameter to do that? Or does the token effectively replace an API key?

Either way, on the back-end, inside an instance: is there a new setting so an index only accepts queries with a tenant token? If not, wouldn't it be fairly easy for someone to bypass restrictions?

Author: Here is the workflow that I hope will help you to understand better:

## User Objective

The user makes a medical SaaS and has an index that contains all medical reports of patients (end-user). He wants to make sure that the end-users will not see the medical reports of other patients.

## Tenant Token to the rescue

1. When the user goes to register or log in, he will have an identifier given by the user's system.
2. This system will generate a token signed by a MeiliSearch API Key (to know at search time that this token has been issued by a "valid" entity (Having knowledge of a MeiliSearch API Key) for this specific end-user. This token will hold within its payload, a description of rules (`indexesPolicy` claim) to be used at search time to ensure that this user can only see his data within a shared index.
3. This token will be given to the frontend code in a transparent way for the end-user.

## Benefits of the solution

- Completely transparent for MeiliSearch. MeiliSearch does not have to store business logic for each end-user and even less to store tokens for each end-user. There could be millions of them.
- The user generates the tenant tokens at will. He does not have to make any specific network exchange with MeiliSearch for that.

```javascript
const tenantTokenRestrictions = {
     "indexesPolicy": {
         "products": {
             "filter": "user_id = 1"
         },
         "reviews": {
             "filter": "user_id = 1 AND published = true"
         }
     },
    "expiresIn": 3600
}

export const generateTenantToken = () => {
  return (parentApiKey: string, restrictions: tenantTokenRestrictions): string => {
    //extract the 8 first characters of the parentApiKey
    const prefix = parentApiKey.substring(0,8);

    //serialize restrictions (indexesPolicies object and expiresIn)
    const queryParameters = serializeQueryParameters(restrictions);

    //create the secured part (parentApiKey + queryParameters)
    const securedKey = crypto
      .createHmac('sha256', parentApiKey)
      .update(queryParameters)
      .digest('hex');

    //return the generated `Tenant Token`
    return Buffer.from(prefix + securedKey + queryParameters).toString('base64');
  };
};
```

##### tenantTokenRestrictions

The format allows defining specific enforced search filters for accessible indexes (these indexes must be defined in the parent `Api Key` used to generate the `Tenant Token` and have the search action).

> If the user does not want to define specific filters for each index accessible to the end-user, he can use the `*` index wildcard rule.

Reviewer: Can we add explicitly that this gives access to all indexes and to all documents only in search? If it's the case.

Author: It does not give access to every index on the MeiliSearch instance but defines this specific filter for every index accessible for the parent API key used to generate the tenant token.

I'll clarify that!

Reviewer: Yeah, sorry I thought that `*` gives access to everything.

Author: No problem, the specification must be more precise.

|--------------------|------------|
| Master Key         | This is the master key that allows the creation of other API keys. The master key is defined by the user when launching MeiliSearch. |
| API Key            | API keys are stored and managed from the endpoint `/keys` by the master key holder. These are the keys used by the technical teams to interact with MeiliSearch at the level of the frontend/backend code. |

> | Tenant Token     | These keys are not stored and managed by a MeiliSearch instance. They are generated for each end-user by the backend code from a MeiliSearch API Key. They are used by the end-users to only search the documents belonging to them. |

Reviewer 3: ```suggestion
| Tenant Token     | These tokens are not stored and managed by a MeiliSearch instance. They are generated for each end-user by the backend code from a MeiliSearch API Key. They are used by the end-users to only search the documents belonging to them. |
```

| Mark    | Mark is a developer for a SaaS company. He will implement the code to communicate with MeiliSearch to solve technical/product needs. |
| UserX   | UserX represents any end-user searching from the frontend interfaces provided by Anna and Mark's company. |

### V. `Tenant Token` Explanations

#### Summary Key Points

> - `Tenant tokens` are generated from a MeiliSearch parent `API key` on the user's backend code to resolve multi-tenancy use-case by restricting access to data within an index according to the criteria chosen by the team managing a MeiliSearch instance.

Reviewer 3: ```suggestion
- `Tenant tokens` are generated from a MeiliSearch parent `API key` on the user's backend code. They are meant to resolve multi-tenancy by restricting access to data within an index according to the criteria chosen by the team managing the MeiliSearch instance.
```

Cut the sentences in chunks for readability

> - These `Tenant tokens` can't be less restrictive than a parent `API key` and can only be used for the search action with a predefined forced filter field.

Reviewer 3: ```suggestion
- `Tenant tokens` cannot be less restrictive than their respective parent `API key` and can only be used for searching with a predefined forced filter field.
```

consistency with previous bullet point

> - These `Tenant tokens` are not stored and thus retrievable on MeiliSearch. This is why we highly advise setting an expiration time on that type of API key for security reasons.

Reviewer 3: ```suggestion
- `Tenant tokens` are not stored and thus not retrievable on MeiliSearch. This is why we highly advise setting an expiration time on them for security reasons.
```

#### Solving Multi-Tenancy with `Tenant tokens`

![](https://i.imgur.com/J4jVe1n.png)

Let's say that `Mark` is a developer for a SaaS platform. He would like to ensure that an end-user can only access their documents at search time. **His database contains many users and he hopes to have many more in the future.**

> When a user registers, the backend code generates a `Tenant token` specifically for that end-user so he can only access his documents.

Reviewer 3: ```suggestion
When a user registers, the backend code generates a `Tenant token` specifically for that end-user so they can only access their documents.
```
